chef/cookbooks/utils/libraries/restart_manager.rb

```diff
@@ -0,0 +1,52 @@
+module ServiceRestart
+  class RestartManager
+    def initialize(cookbook_name, node)
+      @cookbook_name = cookbook_name
+      @node = node
+    end
+
+    def cookbook
+      # if the cookbook_name is apache2, then any cookbook could have triggered it, so
```

Comment:
I guess we'll need a similar approach for haproxy and any other shared services in the future?

Reply:
I don't think so. The problem with apache is that we reload it when, say, keystone.conf is changing. This is not true for haproxy.

Reply:
@vuntz I don't understand you.  Couldn't any cookbook using haproxy could trigger the need for haproxy to restart, if that cookbook changes its usage of haproxy?  Or is your point that that would be done with a reload not a restart?

Reply:
The point is that haproxy is not restarted when keystone.conf/ceilometer.conf/etc. changes.

Let's take concrete examples:

- keystone.conf is changed due to, say, some ldap config change. This would usually trigger a restart/reload from apache from within the keystone cookbook.
- ceilometer.conf is changed due to whatever config change. This would usually trigger a restart/reload from apache from within the ceilometer cookbook.

and:

- haproxy.conf is changed for some change specific to the keystone backend. This would trigger a restart/reload from the crowbar-pacemaker cookbook.
- haproxy.conf is changed for some change specific to the ceilometer backend. This would trigger a restart/reload, still from the crowbar-pacemaker cookbook.

The difference is that apache can be restarted from multiple cookbooks. This is not true for haproxy where the service restart/reload only occurs from one place, even though it is true that the config for haproxy is influenced by several cookbooks.